backend/src/main/scala/cromwell/backend/standard/StandardAsyncExecutionActor.scala

```diff
@@ -362,30 +365,58 @@ trait StandardAsyncExecutionActor extends AsyncBackendJobExecutionActor with Sta
   lazy val runtimeEnvironment = {
     RuntimeEnvironmentBuilder(jobDescriptor.runtimeAttributes, jobPaths)(standardParams.minimumRuntimeSettings) |> runtimeEnvironmentPathMapper
   }
-  
-  lazy val evaluatedAdHocFiles = {
-    val callable = jobDescriptor.taskCall.callable
-    
-    /*
-    NOTE: This method jumps through hoops to keep track of inputs and paths especially for ad hoc files.
-
-    If / when the [[WomFile.value]] is differentiated into cloud paths, VM paths & docker container paths then a lot of
-    this will need to be refactored.
-     */
 
-    def validateAdHocFile(value: WomValue): ErrorOr[List[WomFile]] = value match {
-      case womFile: WomFile => List(womFile).valid
-      case womArray: WomArray => womArray.value.toList.traverse(validateAdHocFile).map(_.flatten)
-      case other =>
-        s"Ad-hoc file creation expression invalidly created a ${other.womType.toDisplayString} result.".invalidNel
+  /**
+    * By default, ad hoc values get localized to the call directory.
+    * This way if running locally with docker they get mounted with the rest of the inputs in the container.
+    * The PAPI backend overrides this to a noop since the localization happens on the VM directly, so there's no need
+    * for this extra localization step.
+    * 
+    * Maybe this should be the other way around: the default implementation is noop and SFS / TES override it ?
+    */
+  lazy val localizeAdHocValues: List[AdHocValue] => ErrorOr[List[StandardAdHocValue]] = { adHocValues => 
+    import cats.instances.future._
+
+    // Localize an adhoc file to the callExecutionRoot as needed
+    val localize: (AdHocValue, Path) => Future[StandardAdHocValue] = {
+      // If the original file is not already under callExecutionRoot, then copy it there
+      case (adHoc @ AdHocValue(_, None, _), file) if !file.isChildOf(jobPaths.callExecutionRoot) =>
+        asyncIo.copyAsync(file, jobPaths.callExecutionRoot / file.name) as {
+          Coproduct[StandardAdHocValue](LocalizedAdHocValue(adHoc, jobPaths.callExecutionRoot / file.name))
+        }
+      // If the file should be available as a different name than its own, copy it as that
+      case (adHoc @ AdHocValue(_, Some(alternateName), _), file) =>
+        val finalPath = jobPaths.callExecutionRoot / alternateName
+        // First check that it's not already there under execution root
+        asyncIo.existsAsync(finalPath) flatMap {
```

Comment:
is there a reason why the previous case uses a blocking `isChildOf` and this does `existsAsync`?

Reply:
Hmm I was assuming that `isChildOf` was purely a path thing to answer the question "is this file located in this path hierarchy ?" and did not involve any I/O. But drilling into it it looks like it's calling `isDirectory` which does involve reading the attributes of the directory and would correspond to an http request for gcs paths.. I would be tempted to say that since the PAPI Backend overrides this to a noop it's likely not going to be an issue for now.
The reason for `isChildOf` is that since we mount the whole call directory in docker, if it's already in there for whatever reason there's no need to re-localize it.